File: sistema.boletas.backend/src/main/java/drej/sistema/boletas/controller/AdminController.java
package drej.sistema.boletas.controller;

import drej.sistema.boletas.config.JwtUtil;
import drej.sistema.boletas.models.Boleta;
import drej.sistema.boletas.models.Usuario;
import drej.sistema.boletas.models.record.BoletaDTO;
import drej.sistema.boletas.models.record.PersonaDTO;
import drej.sistema.boletas.services.BoletaService;
import drej.sistema.boletas.services.PersonaService;
import drej.sistema.boletas.services.UsuarioService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/admin")
public class AdminController {

    @Autowired
    private BoletaService boletaService;

    @Autowired
    private PersonaService personaService;

    @Autowired
    private UsuarioService usuarioService;

    @Autowired
    private JwtUtil jwtUtil;

    // ----------------------- BOLETAS (ADMIN) -----------------------

    // Listar todas las boletas paginadas
    @GetMapping("/boletas")
    public ResponseEntity<Page<BoletaDTO>> listarBoletas(
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "30") int size
    ) {
        Page<Boleta> boletas = boletaService.listarBoletasPaginado(page, size);
        Page<BoletaDTO> dtos = boletas.map(boletaService::toBoletaDTO);
        return ResponseEntity.ok(dtos);
    }

    // Listar boletas de una persona por id
    @GetMapping("/boletas/persona/{personaId}")
    public ResponseEntity<List<BoletaDTO>> listarBoletasPorPersona(@PathVariable Long personaId) {
        List<BoletaDTO> boletas = boletaService.obtenerBoletasID(personaId);
        return ResponseEntity.ok(boletas);
    }

    // Subir/crear varias boletas
    @PostMapping("/boletas")
    public ResponseEntity<?> subirBoletas(@RequestBody List<BoletaDTO> lista) {
        boletaService.guardarBoletas(lista);
        return ResponseEntity.ok().build();
    }

    // Editar boleta
    @PutMapping("/boletas/{id}")
    public ResponseEntity<BoletaDTO> editarBoleta(@PathVariable Long id, @RequestBody BoletaDTO dto) {
        Boleta boleta = boletaService.editarBoleta(id, dto);
        return ResponseEntity.ok(boletaService.toBoletaDTO(boleta));
    }

    // Eliminar boleta
    @DeleteMapping("/boletas/{id}")
    public ResponseEntity<?> eliminarBoleta(@PathVariable Long id) {
        boletaService.eliminarBoleta(id);
        return ResponseEntity.ok().build();
    }

    // ----------------------- PERSONAS (ADMIN) -----------------------

    // Listar todas las personas paginadas (DTO)
    @GetMapping("/personas")
    public ResponseEntity<Page<PersonaDTO>> listarPersonas(
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "30") int size
    ) {
        var pageable = org.springframework.data.domain.PageRequest.of(page, size);
        var personas = personaService.listarPersonasPaginado(pageable);
        return ResponseEntity.ok(personas);
    }

    // Crear persona
    @PostMapping("/personas")
    public ResponseEntity<PersonaDTO> crearPersona(@RequestBody PersonaDTO dto) {
        PersonaDTO nueva = personaService.crearPersona(dto);
        return ResponseEntity.ok(nueva);
    }

    // Editar persona
    @PutMapping("/personas/{id}")
    public ResponseEntity<PersonaDTO> editarPersona(@PathVariable Long id, @RequestBody PersonaDTO dto) {
        PersonaDTO editada = personaService.editarPersona(id, dto);
        return ResponseEntity.ok(editada);
    }

    // Eliminar persona
    @DeleteMapping("/personas/{id}")
    public ResponseEntity<?> eliminarPersona(@PathVariable Long id) {
        personaService.eliminarPersona(id);
        return ResponseEntity.ok().build();
    }

    // ----------------------- USUARIOS (ADMIN) -----------------------

    // Listar todos los usuarios (solo admin)
    @GetMapping("/usuarios")
    public ResponseEntity<?> listarUsuarios(@RequestHeader("Authorization") String token) {
        if (!jwtUtil.isAdmin(token)) {
            return ResponseEntity.status(403).body("No autorizado");
        }
        return ResponseEntity.ok(usuarioService.listarTodos());
    }

    // Cambiar estado de usuario (solo admin)
    @PutMapping("/usuarios/{id}/estado")
    public ResponseEntity<?> cambiarEstado(
            @PathVariable Long id,
            @RequestBody boolean nuevoEstado,
            @RequestHeader("Authorization") String token
    ) {
        if (!jwtUtil.isAdmin(token)) {
            return ResponseEntity.status(403).body("No autorizado");
        }
        Usuario actualizado = usuarioService.actualizarEstado(id, nuevoEstado);
        return ResponseEntity.ok(actualizado);
    }
}

File: backend-boletas-laravel/app/Http/Controllers/AdminController.php
<?php

namespace App\Http\Controllers;

use App\Models\Boleta;
use App\Models\Persona;
use App\Models\Usuario;
use App\Services\BoletaService;
use Illuminate\Http\Request;
use App\Services\UsuarioService;

class AdminController extends Controller
{
    public function __construct(private BoletaService $boletaService) {}

    // --- BOLETAS (ADMIN) ---
    // Listar todas las boletas paginadas
    public function listarBoletas(Request $request)
    {
        $page = (int) $request->get('page', 0);
        $size = (int) $request->get('size', 30);
        $pageData = $this->boletaService->listarBoletasPaginado($page, $size);
        return response()->json($pageData);
    }

    // Listar boletas de una persona por id
    public function listarBoletasPorPersona($personaId)
    {
        $boletas = $this->boletaService->obtenerBoletasPorPersona($personaId);
        return response()->json($boletas);
    }

    // Subir/crear varias boletas
    public function subirBoletas(Request $request)
    {
        $boletas = $request->all();
        $this->boletaService->guardarBoletas($boletas);
        return response()->json(true);
    }

    // Editar boleta
    public function editarBoleta($id, Request $request)
    {
        $boleta = $this->boletaService->editarBoleta($id, $request->all());
        return response()->json($boleta);
    }

    // Eliminar boleta
    public function eliminarBoleta($id)
    {
        $this->boletaService->eliminarBoleta($id);
        return response()->json(true);
    }

    // --- PERSONAS (ADMIN) ---
    public function listarPersonas(Request $request)
    {
        $personas = Persona::all();
        return response()->json([
            'content' => $personas->values()->all(),
            'totalElements' => $personas->count(),
            'totalPages' => 1,
            'size' => $personas->count(),
            'number' => 0,
        ]);
    }

    //crear una persona
    public function crearPersona(Request $request)
    {
        $persona = Persona::create($request->all());
        return response()->json($persona);
    }

    //editar una persona
    public function editarPersona($id, Request $request)
    {
        $persona = Persona::findOrFail($id);
        $persona->update($request->all());
        return response()->json($persona);
    }
    //elimina una persona
    public function eliminarPersona($id)
    {
        Persona::destroy($id);
        return response()->json(true);
    }

    // --- USUARIOS (ADMIN) ---
    public function listarUsuarios()
    {
        $usuarios = \App\Models\Usuario::all()->map(function ($u) {
            return [
                'id' => $u->id,
                'nombre' => $u->nombre,
                'apellido' => $u->apellido,
                'correo' => $u->correo,
                'dni' => $u->dni,
                'telefono' => $u->telefono,
                'rol' => $u->rol,
                'estadoCuenta' => (bool)$u->estado_cuenta,
                'created_at' => $u->created_at,
                'updated_at' => $u->updated_at,
            ];
        });
        return response()->json($usuarios);
    }

    //actualiza el estado de la persona (permite acceder al sistema)
    public function cambiarEstado($id, Request $request)
    {
        $nuevoEstado = $request->json()->all();
        if (is_array($nuevoEstado) && count($nuevoEstado) === 1 && array_key_exists(0, $nuevoEstado)) {
            $nuevoEstado = $nuevoEstado[0];
        } elseif (is_array($nuevoEstado) && isset($nuevoEstado['nuevo_estado'])) {
            $nuevoEstado = $nuevoEstado['nuevo_estado'];
        } elseif (is_array($nuevoEstado) && isset($nuevoEstado['nuevoEstado'])) {
            $nuevoEstado = $nuevoEstado['nuevoEstado'];
        }
        if (!in_array($nuevoEstado, [0, 1, true, false, '0', '1'], true)) {
            return response()->json(['error' => 'Debe enviar un estado booleano'], 422);
        }

        $usuario = $this->usuarioService->actualizarEstado($id, $nuevoEstado);
        return response()->json($usuario);
    }
}